### DISCUSSION TITLE
DisallowConcurrentExecution in a cluster

### DISCUSSION THREAD

**Discussion by cn-byn** [quartznet/quartznet#1641] [Q&A]
Hi,

assuming a cluster setup with two schedulers with the same instance name and different instance ids, is the DisallowConcurrentExecution attribute supposed to prevent the concurrent execution of a job across the whole cluster (i.e., both schedulers) or only "inside" one of the nodes?

So if I add two triggers with exactly the same job key, is the execution of the second trigger supposed to be delayed until the first one has finished in this scenario?

Thanks in advance.

**Comment by lahma** [comment]
Yes, Quartz will block other triggers which target the same job.

https://github.com/quartznet/quartznet/blob/e9d67382cd8c9c60e5b00cb77b329d546ad54319/src/Quartz/Impl/AdoJobStore/JobStoreSupport.cs#L2908-L2922

**Reply by cn-byn** [reply]
Thanks for the quick response.

However, I'm consistenly seeing that a job with the same job key is being executed simultaneously on two different instances (same scheduler). Both instances are running on the same computer, so there isn't an issue with clocks.

For this test, I'm using the following settings.

```json
{
    "quartz.scheduler.instanceId": "...", // "chndefault" and "ServiceConsole_NB043"
    "quartz.scheduler.idleWaitTime": "10000",
    "quartz.scheduler.batchTriggerAcquisitionMaxCount": "1",
    "quartz.threadPool.threadCount": "1",
    "quartz.threadPool.type": "Quartz.Simpl.DefaultThreadPool, Quartz",
    "quartz.jobStore.type": "Quartz.Impl.AdoJobStore.JobStoreTX, Quartz",
    "quartz.jobStore.acquireTriggersWithinLock": "true",
    "quartz.jobStore.misfireThreshold": "60000",
    "quartz.jobStore.driverDelegateType": "Quartz.Impl.AdoJobStore.SqlServerDelegate, Quartz",
    "quartz.jobStore.clustered": "true",
    "quartz.jobStore.dataSource": "db",
    "quartz.dataSource.db.provider": "SqlServer",
    "quartz.dataSource.db.connectionString": "...",
    "quartz.serializer.type": "json"
  }
```

I scheduled 4 simple triggers all with the same job key and with the same time.
For testing purposes, the Job itself just awaits Task.Delay for some time.

Executing the following SQL queries...

```sql
select * from QRTZ_TRIGGERS;
select * from QRTZ_FIRED_TRIGGERS;
```

... yields this:

![image](https://user-images.githubusercontent.com/75720455/175882188-3728264c-19be-4667-9498-aedfe47f6af4.png)

IS_NONCONCURRENT is 1 and the job key is identical, but there are 2 jobs running at the same time on different instances.

Am I missing something?

**Reply by lahma** [reply]
So if we don't look at the implementation detail (database tables), you can see that the two console apps are running the job simultaneously?  I'm just trying to avoid going into a lot of debugging and hair-pulling here...

**Reply by cn-byn** [reply]
Unfortunately, yes.

Here's my TestJob:

```cs
[DisallowConcurrentExecution]
public class TestJob : IJob
{
    public async Task Execute(IJobExecutionContext context)
    {
        Write("BEGIN TestJob.Execute", context);

        for (int i = 0; i < 6; i++)
        {
            Write("Delay 10 s", context);
            await Task.Delay(10000);
        }

        Write("END TestJob.Execute", context);
    }

    private void Write(string msg, IJobExecutionContext context)
    {
        var now = DateTime.UtcNow.ToString("HH:mm:ss.fffffff");
        var scheduler = context.Scheduler.SchedulerName;
        var instance = context.Scheduler.SchedulerInstanceId;
        Debug.WriteLine($"{now} [{scheduler}|{instance}] {msg}");
    }
}
```

With two instances running (both scheduler "Default"), scheduling two identical triggers one minute in the future produces the following debug output:

```
11:18:00.5869851 [Default|chndefault] BEGIN TestJob.Execute
11:18:00.5908448 [Default|chndefault] Delay 10 s
11:18:10.5999037 [Default|chndefault] Delay 10 s
11:18:20.6238008 [Default|chndefault] Delay 10 s
11:18:30.6408646 [Default|chndefault] Delay 10 s
11:18:40.6502410 [Default|chndefault] Delay 10 s
11:18:50.6564777 [Default|chndefault] Delay 10 s
11:19:00.6646496 [Default|chndefault] END TestJob.Execute

11:18:02.4044645 [Default|ServiceConsole_NB043] BEGIN TestJob.Execute
11:18:02.4080580 [Default|ServiceConsole_NB043] Delay 10 s
11:18:12.4186936 [Default|ServiceConsole_NB043] Delay 10 s
11:18:22.4274590 [Default|ServiceConsole_NB043] Delay 10 s
11:18:32.4464402 [Default|ServiceConsole_NB043] Delay 10 s
11:18:42.4682424 [Default|ServiceConsole_NB043] Delay 10 s
11:18:52.4892730 [Default|ServiceConsole_NB043] Delay 10 s
11:19:02.4957294 [Default|ServiceConsole_NB043] END TestJob.Execute
```

Both instances acquire one trigger each, which is fine.
But somehow the functionality that sets other ACQUIRED triggers to BLOCKED before they get executed doesn't work for us (the call to UpdateTriggerStatesForJobFromOtherState in TriggerFired). The second instance still executes its previously acquired trigger.

Maybe it has something to do with isolation levels, but setting txIsolationLevelSerializable to true doesn't help either.

Also, a trigger acquired by the first instance isn't acquired by the second one, so the part that sets a single trigger to ACQUIRED (call to UpdateTriggerStateFromOtherStateWithNextFireTime in AcquireNextTrigger) works, even though the transaction of the second instance may be already waiting for the lock. Once it has the lock, it correctly sees that one trigger has been acquired and acquires the other one.

I think acquiring a non-concurrent job's trigger should immediately block all the others (in the first transaction) to prevent them from being acquired in the first place instead of setting them to BLOCKED in the second transaction. However, that still doesn't explain why the current implementation doesn't work for us.

**Comment by Gavinluo** [comment]
i have the same question,  DisallowConcurrentExecution not work right some time . Quartz 3.3.3

**Comment by aleksandar-angelov** [comment]
We came up with the same scenario where the DisallowConcurrentExecution was not working. The solution was to create the trigger with the same name on every instance., specifying a Triggerkey. If you don't specify a trigger name, Quartz will automatically generate an ID and even though the trigger is the same, the name will be different in every instance.

Here is an example, Quartz version 3.8.1

![image](https://github.com/quartznet/quartznet/assets/151532571/6226d01d-37b8-4882-b104-bd228916bf72)

**Reply by cn-byn** [reply]
Thanks, but that's not really a solution since now you can only have a single trigger (with that specific TriggerKey).

DisallowConcurrentExecution is meant to work w.r.t. the JobKey, though. So the number of triggers and their names should not matter.

But maybe that's a hint why the unexpected behavior occurs?

**Reply by aleksandar-angelov** [reply]
Basically, in your scenario, you have the same trigger that will occur at some time in both instances, so both instances will need to take care of the same trigger and the same job. IMO, there is no need to have multiple triggers that have the same schedule and same job. You need to have one trigger and one of the cluster nodes, randomly chosen, will execute the trigger with the job and will prevent a concurrent execution if the next firing schedule came from another cluster node.

**Reply by cn-byn** [reply]
> IMO, there is no need to have multiple triggers that have the same schedule and same job.

But there is this need in my case. I have several simple triggers (i.e., non-recurring triggers) and those should not cause the job to run concurrently if DisallowConcurrentExecution is set.

Besides, this also happens with multiple triggers with different schedules.

**Comment by DominicUllmann** [comment]
Sadly, I also see this issue still in 3.13.0.

To test this, I use a setup with 2 nodes running on the same machine. 
I have multiple non-recurring triggers for the same job-key. If the trigger time is around the same time, it is possible that multiple triggers fire on different instances at the same time running the job at the same time although the job has DisallowConcurrentExecution attribute set.

To make the issue more likely to happen, I have in my test scheduled the triggers around the same time and simulated a long run time of the job itself (similar to the initial example).
I use the following config setup for the jobs:

            var job = JobBuilder.Create<TestJob>()
                .WithIdentity("job1", "group1")
                .Build();

            var timeToStart = DateTime.Now + TimeSpan.FromSeconds(15);
            
            var trigger2a = TriggerBuilder.Create()
                .WithIdentity("trigger2_s", "group1")
                .ForJob(job)
                .StartAt(timeToStart + TimeSpan.FromSeconds(0)).Build();
            await scheduler.ScheduleJob(job, trigger2a);
            
            for (int i = 0; i < 10; i++)
            {
                var trigger2 = TriggerBuilder.Create()
                    .WithIdentity("trigger2_" + i , "group1")
                    .ForJob(job)
                    .StartAt(timeToStart + TimeSpan.FromSeconds(Math.Min(i, 4))).Build();
                await scheduler.ScheduleJob(trigger2);
            }
        }

        await scheduler.Start()`

Based on this setup, I see that the triggers run at the beginning on different instances at the same time. Later on, only 1 instances continues probably related to the misfiring.

Initially I see that 2 triggers are aquired at the same time:

![image](https://github.com/user-attachments/assets/e4f4c714-bb99-473d-871b-06b24b8264ba)

When the jobs are running, I see:

![image](https://github.com/user-attachments/assets/d17baef8-fa4b-4ba0-8e7d-f63d5a376210)

The job looks like this:

![image](https://github.com/user-attachments/assets/3d7d4173-12ea-441a-bc52-f5e0e0fb85ae)

I even once managed, that on the second node, 2 triggers were aquired at the same time running the job twice on the same node:

![image](https://github.com/user-attachments/assets/4c3c9f40-cea1-442f-b6d7-18914f7a0bb4)

The job itself is very simple:

`[DisallowConcurrentExecution]
public class TestJob : IJob
{
    public async Task Execute(IJobExecutionContext context)
    {
        Write("BEGIN TestJob.Execute", context);
     
             for (int i = 0; i < 4; i++)
             {
                 Write("Delay 10 s", context);
                 await Task.Delay(10000);
             }
     
             Write("END TestJob.Execute", context);
         }
     
         private void Write(string msg, IJobExecutionContext context)
         {
             var now = DateTime.UtcNow.ToString("HH:mm:ss.fffffff");
             var scheduler = context.Scheduler.SchedulerName;
             var instance = context.Scheduler.SchedulerInstanceId;
             Console.WriteLine($"{now} [{scheduler}|{instance}|{context.Trigger.Key}] {msg}");
         }
}`

Normally, the output looks like this:
Node1:
![image](https://github.com/user-attachments/assets/41c66472-cb29-400c-afa2-be17a17d503b)
Node2:
![image](https://github.com/user-attachments/assets/bba11d0c-8c4c-45d0-a813-5d5bcb32ff10)


Any ideas why that is happening? What can I do to have the DisallowConcurrentExecution reliably preventing concurrent runs?

Thanks.

**Reply by AndreasKorn** [reply]
We are seeing the same issue with multiple nodes. The job runs are not blocked and run concurrently.

**Comment by pigiax** [comment]
I think I get the issue!
To run a job there are two steps:

1. select (or acquire) of the trigger
2. lock of other triggers with same job_name on same scheduler.

The first step is done here:
https://github.com/quartznet/quartznet/blob/e9d67382cd8c9c60e5b00cb77b329d546ad54319/src/Quartz/Core/QuartzSchedulerThread.cs#L283

The second step is done here:
https://github.com/quartznet/quartznet/blob/e9d67382cd8c9c60e5b00cb77b329d546ad54319/src/Quartz/Core/QuartzSchedulerThread.cs#L381

Both steps individually are granted to run exclusively with a database lock, but in between of them there are many rows of code which are not synchronized, so the concurrency can occur very easily.
Ideally the acquire and the lock of the other triggers should happen into the same transaction, with a db lock, to grant that, in a clustered environment, a single execution happens.

Now, as workaround, I've written a custom `IJobStore` based on the Quartz `JobStoreTX` which does the acquire and the lock of other triggers in the same transaction:

```
public class CustomJobStore : JobStoreTX
{
    protected override async Task<IReadOnlyCollection<IOperableTrigger>> AcquireNextTrigger(ConnectionAndTransactionHolder conn, DateTimeOffset noLaterThan, int maxCount, TimeSpan timeWindow, CancellationToken cancellationToken = default)
    {
        var triggers = await base.AcquireNextTrigger(conn, noLaterThan, maxCount, timeWindow, cancellationToken);

        if (triggers != null && triggers.Count > 0)
        {
            foreach (IOperableTrigger trigger in triggers)
            {
                try
                {
                    await Delegate.UpdateTriggerStatesForJobFromOtherState(conn, trigger.JobKey, StateBlocked, StateWaiting, cancellationToken).ConfigureAwait(false);
                    await Delegate.UpdateTriggerStatesForJobFromOtherState(conn, trigger.JobKey, StatePausedBlocked, StatePaused, cancellationToken).ConfigureAwait(false);
                }
                catch (Exception e)
                {
                    throw new JobPersistenceException("Couldn't update states of blocked triggers: " + e.Message, e);
                }
            }
        }
        return triggers;
    }
}
```

Then it's only needed to override the properties of the scheduler:
```
properties["quartz.jobStore.type"] = "MyProject.Quartz.CustomJobStore, CustomJobStore";
```

Now I need to test it deeply in a very intensive scenario, but seems to works for the moment.
Tell me if I have said something wrong @lahma

**Reply by pigiax** [reply]
I created an issue to track this problem: #2697
